### PR TITLE
Fixes #8178, 8189:  adding UI for incremental update.

### PR DESCRIPTION
**Pull request by waldenraines**
http://projects.theforeman.org/issues/8178
http://projects.theforeman.org/issues/8189

### CODE REVIEW

**Comment by waldenraines** [comment]
Wanted to go ahead and open this so I could start getting feedback.  Still need to address the following:

- [ ] Fix existing failing tests
- [ ] Add tests for new functionality

as well as any comments y'all have.

**Comment by ehelms** [comment]
@waldenraines Would you mind posting workflows/scenarios to test for this PR?

**Comment by waldenraines** [comment]
@ehelms https://dl.dropboxusercontent.com/u/5892944/Emergency Errata Management-2014-11-14.pdf

**Comment by ehelms** [comment]
This is what I got when I clicked 'Apply to Hosts':

apply errata

**Comment by waldenraines** [comment]
@ehelms the screenshot above is for the case when there is no incremental update needed (i.e. the errata is available in the content host's cv/env.).  In order to see the other screen you'd have to filter out an errata and promote the view and then update the host to use the promoted view.  This screen isn't documented in the mockups.

**Comment by ehelms** [comment]
@waldenraines Should the above screen be the errata ID and not the UUID?

**Comment by waldenraines** [comment]
> @waldenraines Should the above screen be the errata ID and not the UUID?

Yep, will fix.

**Comment by ehelms** [comment]
1. Seeing the same errata id issue on the apply screen w/ the incremental update.

**Comment by ehelms** [comment]
After visinting the task, I went backwards in my browser to return to the errata > content hosts tab and was presented wtih:

apply errata2

This seems like the wrong output to be showing me given I can't select anything and this message is shortened and looks strange as opposed to fitting across the entire length.

**Comment by ehelms** [comment]
Not sure how related to your changes this is, but my attempt to create the incremental update resulted in:

```
Katello::Errors::PulpError: PLP0000: Could not find a published directory for Default_Organization-ZooView-4_1-Zoo-Fake_Repo.
```

**Comment by ehelms** [comment]
I tried to publish another incremental update, the server ISE in the background but the UI just sat there at the confirm page telling me nothing.

**Comment by ehelms** [comment]
So I created 4.0 and 4.1, where 4.0 was in Dev and got incrementally updated to 4.1. I then went to promote 4.0 to Library since prior to this, 3.0 was in Library. When I did I got the following:

promote content view zooview

**Comment by ehelms** [comment]
Choosing to apply the change while generating the incremental update appeared to do nothing for me,

**Comment by waldenraines** [comment]
1.  
> Not sure how related to your changes this is, but my attempt to create the incremental update resulted in:
> 
> Katello::Errors::PulpError: PLP0000: Could not find a published directory for Default_Organization-ZooView-4_1-Zoo-Fake_Repo.

2.  
> I tried to publish another incremental update, the server ISE in the background but the UI just sat there at the confirm page telling me nothing. Note: I resumed the task and it completed successfully. This behavior happens consistently - the successful resume, though, does include the errata needed by the content host.

3.  
> So I created 4.0 and 4.1, where 4.0 was in Dev and got incrementally updated to 4.1. I then went to promote 4.0 to Library since prior to this, 3.0 was in Library. When I did I got the following: "Cannot promote environment out of sequence.  Use force to bypass restriction".

@jlsherrill any thoughts on these?  Obviously for the second quote I need to present the error to the user but the fact that the tasks keep failing may be indicative of an issue server side.

> Choosing to apply the change while generating the incremental update appeared to do nothing for me,

That is not yet implemented.  See http://projects.theforeman.org/issues/8191 and http://projects.theforeman.org/issues/8178

**Comment by jlsherrill** [comment]
@waldenraines i hit that while working on https://github.com/Katello/katello/pull/4883

and i believe it is fixed in that PR here:  https://github.com/Katello/katello/pull/4883/files#diff-fa7ec1c4e8306f6de501153314ba7ee3R19

Its basically a race condition that only sometimes happens.  It probably is more consistent on larger (or smaller repos).

**Comment by ehelms** [comment]
@waldenraines I'll re-review once the PR @jlsherrill mentioned is merged and this is re-based. As for the applying the update - are you planning to await that API change for this PR or hide it so that this part of the work can go in sooner?

**Comment by waldenraines** [comment]
> @waldenraines I'll re-review once the PR @jlsherrill mentioned is merged and this is re-based. As for the applying the update - are you planning to await that API change for this PR or hide it so that this part of the work can go in sooner?

That's a good point.  I should probably just remove that checkbox for now and then add it after the apply after update is implemented.

**Comment by jlsherrill** [comment]
errata-2

Any particular reason this is so tight horizontally?

**Comment by jlsherrill** [comment]
* when bulk selecting errata and clicking 'apply errata', the selected errata are no longer selected on the left hand side.  Closing that pane shows them selected again
* Possibly related, but the call to /katello/api/v2/systems? to fetch the list of systems is being made with these params:
```
organization_id:1
page:1
search:
```

none of these params include the selected errata.

**Comment by waldenraines** [comment]
> Any particular reason this is so tight horizontally?

Not really, I can add a `.details-full` to make that full width of the details pane.

**Comment by waldenraines** [comment]
@jlsherrill @ehelms updated.  I haven't been able to figure out this issue yet:

> when bulk selecting errata and clicking 'apply errata', the selected errata are no longer selected on the left hand side. Closing that pane shows them selected again

But I'm still looking into it and believe that we can fix it with a follow up bug fix if necessary.

**Comment by waldenraines** [comment]
> But I'm still looking into it and believe that we can fix it with a follow up bug fix if necessary.

I figured out the issue right after I typed this.  Updated.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    <p translate ng-hide="errata">
+      Apply {{ errataIds.length }} Errata to {{ selectedContentHosts.included.ids.length }} Content Host(s)?
+    </p>
+  </section>
+
+  <section ng-show="updates">
+    <p translate>
+      These Errata were not available in your published Content View Versions running on the selected hosts.  The new Content View Versions (specified below)
+      will be created which will make this Errata available in the host's Environment.  This new version will replace the current version in your host's Lifecycle
+      Environment.  To install these errata immediately on hosts after publishing check the box below.
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html]: I see no box below to check.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    </p>
+    <p translate ng-hide="errata">
+      Apply {{ errataIds.length }} Errata to {{ selectedContentHosts.included.ids.length }} Content Host(s)?
+    </p>
+  </section>
+
+  <section ng-show="updates">
+    <p translate>
+      These Errata were not available in your published Content View Versions running on the selected hosts.  The new Content View Versions (specified below)
+      will be created which will make this Errata available in the host's Environment.  This new version will replace the current version in your host's Lifecycle
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html]: should this be `host's Lifecycle Environment` ?

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      Apply {{ errata.errata_id }} to {{ selectedContentHosts.included.ids.length }} Content Host(s)?
+    </p>
+    <p translate ng-hide="errata">
+      Apply {{ errataIds.length }} Errata to {{ selectedContentHosts.included.ids.length }} Content Host(s)?
+    </p>
+  </section>
+
+  <section ng-show="updates">
+    <p translate>
+      These Errata were not available in your published Content View Versions running on the selected hosts.  The new Content View Versions (specified below)
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html]: Wondering if we should capitalize 'versions'

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          </td>
+          <td>
+            {{ update.content_host_count }}
+          </td>
+        </tr>
+      </tbody>
+    </table>
+  </section>
+
+  <form ng-submit="confirmApply()">
```
Comment (ehelms) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html]: For me, these buttons are up tight on the table.

**Comment by ehelms** [comment]
Confirm button showed no indication an action was taking place in the background, and I am not the biggest fan of the user having to be taken out of context over to the tasks page after starting it.

**Comment by ehelms** [comment]
When I click 'Back' from the tasks page I am presented with the errata apply page and spinners.

**Comment by ehelms** [comment]
After applying an incremental update to my content view, and looking at the errata list, it tells me that I have 1 applicable, 1 installable in the content host list. But I only have 1 content host?

**Comment by ehelms** [comment]
For bulk incremental update, on the 'Next' button we show 1 of 2, but never show 2 of 2. We've been generally showing breadcrumbs when there is a workflow or set of nested states. 

I see the following on the bulk errata apply page:
```
 TypeError: Cannot read property 'hasOwnProperty' of undefined
    at Object.fn (page-title.directive.js?body=1:46)
    at Scope.$get.Scope.$digest (angular.js?body=1:11809)
    at Scope.$get.Scope.$apply (angular.js?body=1:12062)
    at HTMLButtonElement.<anonymous> (angular.js?body=1:17840)
    at HTMLButtonElement.jQuery.event.dispatch (jquery.js?body=1:3333)
    at HTMLButtonElement.jQuery.event.add.elemData.handle.eventHandle (jquery.js?body=1:2942)
```

**Comment by ehelms** [comment]
If I start an incremental update, then go back to the errata, the content host is still shown as applicable, and it allows me to go through the incremental update process again with no indication that there is one in progress for that content view to that environment that would affect that host. I could perceive that it didn't work and either keep trying and create a lot of CVs or accidentally file a bug.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html (hunk 1 above)
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html]: Ah, that's leftover from when I had the checkbox you commented on last time still there.  Will update.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html (hunk 2 above)
Comment (waldenraines) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/errata/views/apply-errata-confirm.html]: It is, look at the next line.

**Comment by waldenraines** [comment]
> Confirm button showed no indication an action was taking place in the background, and I am not the biggest fan of the user having to be taken out of context over to the tasks page after starting it.

That was [your suggestion](https://github.com/Katello/katello/pull/4931#issuecomment-70325259)...What would you like to see here instead?


> After applying an incremental update to my content view, and looking at the errata list, it tells me that I have 1 applicable, 1 installable in the content host list. But I only have 1 content host?

It's because these terms are not mutually exclusive.

> For bulk incremental update, on the 'Next' button we show 1 of 2, but never show 2 of 2. We've been generally showing breadcrumbs when there is a workflow or set of nested states.

Just following the [mockups](https://dl.dropboxusercontent.com/u/5892944/Emergency%20Errata%20Management-2014-11-14.pdf).  I can convert to the breadcrumbs if Kyle agrees.

**Comment by waldenraines** [comment]
> If I start an incremental update, then go back to the errata, the content host is still shown as applicable, and it allows me to go through the incremental update process again with no indication that there is one in progress for that content view to that environment that would affect that host. I could perceive that it didn't work and either keep trying and create a lot of CVs or accidentally file a bug.

Can we handle this as an enhancement after this is merged?  I'm happy to file an issue and tackle it separately as this will require some thought.

**Comment by ehelms** [comment]
On Thu, Jan 29, 2015 at 1:25 PM, Walden Raines <notifications@github.com>
wrote:

> Confirm button showed no indication an action was taking place in the
> background, and I am not the biggest fan of the user having to be taken out
> of context over to the tasks page after starting it.
>
> That was your suggestion
> <https://github.com/Katello/katello/pull/4931#issuecomment-70325259>...What
> would you like to see here instead?
>
Yea, I was just commenting on how, as a user it felt to me. I won't block
the PR on it, but I think in general the way we do repository sync by
showing a page in the details pane that shows the task provides a better
experience.

>  After applying an incremental update to my content view, and looking at
> the errata list, it tells me that I have 1 applicable, 1 installable in the
> content host list. But I only have 1 content host?
>
> It's because these terms are not mutually exclusive.
>
> For bulk incremental update, on the 'Next' button we show 1 of 2, but
> never show 2 of 2. We've been generally showing breadcrumbs when there is a
> workflow or set of nested states.
>
> Just following the mockups
> <https://dl.dropboxusercontent.com/u/5892944/Emergency%20Errata%20Management-2014-11-14.pdf>.
> I can convert to the breadcrumbs if Kyle agrees.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Katello/katello/pull/4931#issuecomment-72076828>.
>



-- 
Go Pack and War Eagle,
Eric D. Helms

Ph.D. Student - North Carolina State University

**Comment by ehelms** [comment]
On Thu, Jan 29, 2015 at 1:27 PM, Walden Raines <notifications@github.com>
wrote:

> If I start an incremental update, then go back to the errata, the content
> host is still shown as applicable, and it allows me to go through the
> incremental update process again with no indication that there is one in
> progress for that content view to that environment that would affect that
> host. I could perceive that it didn't work and either keep trying and
> create a lot of CVs or accidentally file a bug.
>
> Can we handle this as an enhancement after this is merged? I'm happy to
> file an issue and tackle it separately as this will require some thought.
>
Not sure I would call it an enhancement since the behavior here can lead a
user down a bad or confusing path that will ultimately lead us to fielding
the same question over and over. However, since you filed an issue already
I am willing to let this PR go without it. I would like to see some thought
and a fix for it prioritized.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Katello/katello/pull/4931#issuecomment-72077112>.
>



-- 
Go Pack and War Eagle,
Eric D. Helms

Ph.D. Student - North Carolina State University

**Comment by waldenraines** [comment]
> Yea, I was just commenting on how, as a user it felt to me. I won't block
> the PR on it, but I think in general the way we do repository sync by
> showing a page in the details pane that shows the task provides a better
> experience.

What about an issue for this as well?

http://projects.theforeman.org/issues/9176

**Comment by ehelms** [comment]
ACK

**Pull request merged**